### DISCUSSION TITLE
What makes the best model?

### DISCUSSION THREAD

**Discussion by terryfrankcombe** [mir-group/nequip#276] [General]
Hi folks

I'm training a model on an (admittedly not very large) dataset.  Frequently I get in the output e.g.
`! Best model       16    0.031`
but only early in the run.  In this case the latest one printed is indeed this 16th epoch result.

The training goes on for some time.  Eventually at epoch 926 I get
`! Stop training: Early stopping: LR is smaller than 1e-05`
Now the loss here is larger, at 0.0426.  However, e_mae is much smaller than it was (0.000877/0.00314 for training/validation, compared to 0.00946/0.00601 at "best"), making me thing that the last model is a better model than the "best".

This is borne out somewhat by nequip-evaluate, which reports e_mae & f_mae to be 0.005180 and 0.006929 for best_model.pth on the evaluation set, but that's reduced to 0.001576 and 0.004805 for last_model.pth.

In what way is this "best" model actually best, and why wouldn't the last model be considered best?

**Comment by Linux-cpp-lisp** [comment]
Hi @terryfrankcombe ,

 `nequip` chooses the best model (and uses the `ReduceLROnPlateau` scheduler) on the field specified by `metrics_key`, which is `validation_loss` (not one of the MAEs) by default. Note that this loss has whatever normalization the training loss has, so if you are using `PerAtom` energy loss that per-atom scaling will apply when computing `validation_loss` as well.  If you logged to wandb (or just plot one of the metrics CSV files) you can confirm this; if your run is also inconsistent with this explanation then we might have a bug...

If you want to schedule on a different metric, you can specify `metrics_key: validation_e_mae` (for example) and the best model (as well as the plateau LR scheduler, if used) will schedule off of `validation_e_mae`.

**Reply by terryfrankcombe** [reply]
OK, that's good information.

If I switch to using e.g. validation_e_mae, does the input force data continue to play a significant role in the training?

**Reply by Linux-cpp-lisp** [reply]
Yes, this change only affects how the learning rate scheduler and best model behave; the training will still include whatever you specified in `loss_coeffs`.
